### PR TITLE
[JIT] Remove extra bound checks for explicit range checks

### PR DESCRIPTION
**Pull request by ArtBlnd**
Original PR https://github.com/dotnet/coreclr/pull/22848 and rebased because of `optRemoveRangeCheck`

Currently, removing bound checks for explicit array(includes buffer, char[], string, array like etcs.) range check on some major patterns.

such as
```
if (array.Length > 0)
{
    // Some (array) use expressions.
}
```

The explicit range check removal will be triggered if only the code scope of if expression is only can reach with if expressions, which means there is no Side-Effect on that code scope.

Related to #19620

### CODE REVIEW

**Comment by ArtBlnd** [comment]
I have no idea with `unexpected type name 'allocator': expected expression` error. someone can help me to fix it out?

**Comment by ArtBlnd** [comment]
image

Here is jit-diff for frameworks.

**Comment by ArtBlnd** [comment]
As I wrote on the original PR. I'll upload testcases next week.

And also, ready for review.@dotnet/jit-contrib

Thanks.

**Comment by ArtBlnd** [comment]
@AndyAyersMS

File: src/jit/rangecheck.cpp [hunk 1]
```diff
@@ -190,18 +190,65 @@ bool RangeCheck::BetweenBounds(Range& range, int lower, GenTree* upper)
...
+            break;
+        }
+        case GT_JTRUE:
+        {
+            GenTree*   gtCmp    = treeParent->gtGetOp1();
+            genTreeOps opCmp    = gtCmp->OperGet();
+            GenTree*   gtCmpLen = gtCmp->gtGetOp1();
+            GenTree*   gtCmpCns = gtCmp->gtGetOp2IfPresent();
+
+            // The compare expression can be comma because of CSE expansion. extract effeictive value before checking it
```
Comment (MarcoRossignoli) [inline src/jit/rangecheck.cpp]: nit: typos `effective`

Comment (ArtBlnd) [reply src/jit/rangecheck.cpp]: Thanks :)

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

**Pull request closed without merging**